include/swift/ABI/TaskGroup.h

```diff
@@ -0,0 +1,393 @@
+        /// when the task was enqueued into the ready queue to keep it alive
+        /// until a next() call eventually picks it up.
+        AsyncTask *retainedTask;
+
+        bool isStorageAccessible() {
+          return status == PollStatus::Success ||
+              status == PollStatus::Error ||
+              status == PollStatus::Empty;
+        }
+
+        static PollResult get(AsyncTask *asyncTask, bool hadErrorResult) {
+          auto fragment = asyncTask->futureFragment();
+          return PollResult{
+              /*status*/ hadErrorResult ?
+                  TaskGroup::PollStatus::Error :
+                  TaskGroup::PollStatus::Success,
+              /*storage*/ hadErrorResult ?
+                  reinterpret_cast<OpaqueValue *>(fragment->getError()) :
+                  fragment->getStoragePtr(),
+              /*task*/ asyncTask
+          };
+        }
+    };
+
+    /// An item within the message queue of a group.
+    struct ReadyQueueItem {
+        /// Mask used for the low status bits in a message queue item.
+        static const uintptr_t statusMask = 0x03;
+
+        uintptr_t storage;
+
+        ReadyStatus getStatus() const {
+          return static_cast<ReadyStatus>(storage & statusMask);
+        }
+
+        AsyncTask *getTask() const {
+          return reinterpret_cast<AsyncTask *>(storage & ~statusMask);
+        }
+
+        static ReadyQueueItem get(ReadyStatus status, AsyncTask *task) {
+          assert(task == nullptr || task->isFuture());
+          return ReadyQueueItem{
+              reinterpret_cast<uintptr_t>(task) | static_cast<uintptr_t>(status)};
+        }
+    };
+
+    /// An item within the pending queue.
+    struct PendingQueueItem {
+        uintptr_t storage;
+
+        AsyncTask *getTask() const {
+          return reinterpret_cast<AsyncTask *>(storage);
+        }
+
+        static ReadyQueueItem get(AsyncTask *task) {
+          assert(task == nullptr || task->isFuture());
+          return ReadyQueueItem{ reinterpret_cast<uintptr_t>(task) };
+        }
+    };
+
+    struct GroupStatus {
+        static const uint64_t cancelled      = 0b1000000000000000000000000000000000000000000000000000000000000000;
+        static const uint64_t waiting        = 0b0100000000000000000000000000000000000000000000000000000000000000;
+
+        // 31 bits for ready tasks counter
+        static const uint64_t maskReady      = 0b0011111111111111111111111111111110000000000000000000000000000000;
+        static const uint64_t oneReadyTask   = 0b0000000000000000000000000000000010000000000000000000000000000000;
+
+        // 31 bits for pending tasks counter
+        static const uint64_t maskPending    = 0b0000000000000000000000000000000001111111111111111111111111111111;
+        static const uint64_t onePendingTask = 0b0000000000000000000000000000000000000000000000000000000000000001;
+
+        uint64_t status;
+
+        bool isCancelled() {
+          return (status & cancelled) > 0;
+        }
```

Comment:
We reject such attempts to add tasks.

@DougGregor's rewording of the proposal make the `add` operation return a `bool`: https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md#adding-tasks-to-a-group

> The add operation returns true to indicate that the task was added to the group, and false otherwise. The only circumstance under which the task will not be added to the group is when the task group has been cancelled; see the section on task group cancellation for more information.

Previously this used to return a Handle to the task; so there wasn't any feedback if it was added or not; though we could do this as well with the handle API, where we would return an immediately failed handle?

Reply:
The rejecting if the group was cancelled is implemented like this:

```swift
    @discardableResult
    public mutating func add(
      overridingPriority priorityOverride: Priority? = nil,
      operation: @concurrent @escaping () async throws -> TaskResult
    ) async -> Bool {
      let canAdd = _taskGroupAddPendingTask(group: _group)

      guard canAdd else {
        // the group is cancelled and is not accepting any new work
        return false
      }
```

where the first operation is an atomic op on the status, in which we also store the cancelled bit, so it is very cheap to know if we're cancelled. We always need to do this add pending (+1 to pending count) anyway, so it is not extra work.